keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,362 @@
+
+### Goals
+
+- Pod Connectivity: IPv4-to-IPv4 and IPv6-to-IPv6 access between pods
+- Service Access: IPv4-to-IPv4 and IPv6-to-IPv6 access from pods to Kubernetes services
+- External Server Access: IPv4-to-IPv4 and IPv6-to-IPv6 access from pods to external servers
+- Ingress Access: Access from IPv4 and/or IPv6 clients to Kubernetes services. Depending on the ingress controller used, connections may cross IP families (IPv4 clients to IPv6 services, or vice versa).
+- Service Discovery: Support for both A and AAAA records for kube-DNS and coreDNS
+- Functionality tested with the Bridge CNI plugin and Host-Local IPAM plugin as references
+- Maintain backwards-compatible support for IPv4-only and IPv6-only clusters
+
+### Non-Goals
+
+- Cross-family connectivity: IPv4-to-IPv6 and IPv6-to-IPv4 connectivity is considered outside of the scope of this proposal, except in the case of ingress access (depending on ingress controller used).
+- CNI network plugins other than the Bridge and Host-Local IPAM plugins should support Kubernetes dual stack, but the development and testing of dual stack support for these other plugins is considered outside of the scope of this proposal.
+- Code changes should support any number IPv4 addresses and any number of IPv6 addresses per pod (up to some reasonable limit). However, testing will be done only up to the following limits:
+  - Pod addresses: 1 IPv4 address and 2 IPv6 addresses per pod maximum
+  - Service addresses: 1 IPv4 and 1 IPv6 service addresses per service maximum
+- For simplicity, for the first release of dual-stack support, the [Kubernetes Endpoints API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#endpoints-v1-core) will not be modified. Each IPv4/IPv6 address that is assigned to a backend pod will be treated and displayed as a separate endpoint for the corresponding service. Service load balancing (e.g. using kube-proxy) will be performed on a per-address basis, rather than being done on a per-dual-stack-endpoint basis.
+
+## Proposal
+
+In order to support dual-stack in Kubernetes clusters, Kubernetes needs to have awareness of and support dual-stack addresses for pods, nodes and services. To make this happen, the following changes will be required:
+
+- Kubernetes needs to be made aware of multiple IPs per pod (tested with up to one IPv4 and up to two IPv6 addresses).
+- Link Local Addresses (LLAs) on a pod will remain implicit (Kubernetes will not display nor track these addresses).
+- Kubernetes needs to be configurable for up to two service CIDRs.
+- Backend pods for a service can be dual stack. For the first release of dual-stack support, each IPv4/IPv6 address of a backend pod will be treated as a separate Kubernetes endpoint.
+- Kube-proxy needs to support IPv4 and IPv6 services in parallel (e.g. drive iptables and ip6tables in parallel).
```

Comment:
We should also consider the impact to the IPVS proxy. I think at this point we need to maintain both (unless we say dual-stack is iptables only?)

Reply:
Thanks for looking into this, it's good to know up front!